py/builtin.h

```diff
@@ -64,7 +64,15 @@ MP_DECLARE_CONST_FUN_OBJ_KW(mp_builtin_open_obj);
 
 #endif
 
+// A port can provide its own import handler to expand or fully replace
+// the mp_builtin___import___default function.
+mp_obj_t mp_builtin___import___default(size_t n_args, const mp_obj_t *args);
+#if MICROPY_MODULE_OVERRIDE_BUILTIN_IMPORT
 mp_obj_t mp_builtin___import__(size_t n_args, const mp_obj_t *args);
+#else
+#define mp_builtin___import__ mp_builtin___import___default
```

Comment:
please use a static inline definition here, it's more robust than a #define

Reply:
Done and rebased. It does increase the code size by a few bytes (+4 on minimal embedded port).

Reply:
Ah!  So the reason is because of this line:
```c
MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mp_builtin___import___obj, 1, 5, mp_builtin___import__);
```
which forces both the default and static-inline wrapper function to be linked in to the firmware (instead of just the default one).

I guess we need to go back to using the `#define` to make it efficient...

Reply:
Thanks! I was wondering about that. I reverted it and rebased.

Reply:
Sorry to go back and forth on this.... We don't really have a consistent way to override an arbitrary function in the core, but there is some precedence set by `py/mphal.h`.  Following the way it does things, we could do the following here:
```c
#ifndef mp_builtin___import__
#define mp_builtin___import__ mp_builtin___import___default
#endif
```
That would be pretty clean, and doesn't require an entry in `py/mpconfig.h`.  It's also how `ports/stm32/boardctrl.h` does things.

Then in your port you do:
```c
mp_obj_t custom_import(size_t n_args, const mp_obj_t *args);
#define mp_builtin___import__ custom_import
```

Reply:
That's one of the first approaches I tried, but I got a bit stuck on where to do this exactly.

I'm able to make this work following the same `mphal` approach (the `ifndef` in `py/mphal.h` and `custom_import` in `mphalport.h`), but I figured this one doesn't belong there since it doesn't concern hardware.

It seems like the custom implementation would be provided via `mpconfigport.h`, but I'm not sure how since the declaration relies on `mp_obj_t` which isn't defined at that point.

What would be the right places to put the suggested snippets above?

Reply:
It should go here in `py/builtin.h`.

How about this:
```c
#ifndef mp_builtin___import__
#define mp_builtin___import__ mp_builtin___import___default
#endif
mp_obj_t mp_builtin___import__(size_t n_args, const mp_obj_t *args);
mp_obj_t mp_builtin___import__default(size_t n_args, const mp_obj_t *args);
```
?

It's a bit tricky but should work.

Reply:
Nice, that works! Updated and rebased.